target/scripts/helper-functions.sh

```diff
@@ -21,7 +21,7 @@ function create_lock
     LOCK_FILE="/tmp/docker-mailserver/${SCRIPT_NAME}.lock"
     [[ -e "${LOCK_FILE}" ]] && errex "Lock file ${LOCK_FILE} exists. Another $1 execution is happening. Try again later."
     trap remove_lock EXIT # This won't work if the script is, for example, check-for-changes.sh which uses a while loop to stay running; you'll need to include a remove_lock call at the end of your logic
-    touch "${LOCK_FILE}"
+    touch "${LOCK_FILE}" &>/dev/null
```

Comment:
Yep, throws this annoying message:
`touch: setting times of '/tmp/docker-mailserver/check-for-changes.sh.lock': Stale file handle`

Reply:
Ok, I understand. But that way, you also won't notice, if the LOCK_FILE can't be created at all? E.g. if someone mounts `/tmp/docker-mailserver` in read-only mode.

Reply:
BTW: Is there need for the lockfile to be in the mounted path (and therefore can survive container restarts)?  Would the /tmp location maybe a better place?

Reply:
Yeah, that's a good point! I've added a check to ensure it gets created. The lockfile must be in a shared location (like my EFS volume) so that every container can see the lock file and prevent making changes all at once. I use a single EFS across multiple regions in AWS for example and there is a chance that the server I'm hitting when I use a load balancer is not the same between my executed commands.

Reply:
One last thing: It could also be worth, on container start (TBH I don't know if that fits your setup), to check for stale lock files? In a simple DMS setup, there shouldn't be any lock files on start.

Reply:
:) I actually have had that issue where I've shut down a container forcefully before and left a stale lock file which had to then be manually deleted. 

Unfortunately, forcing a delete when I start a new container that shares the mount is a bad idea as it could mess up an active/legit lock file that exists due to changes somewhere else. I'd rather it throw the "lock file already exists" warning in the logs until the user notices that none of their changes are making it through. This is a much less risky situation as the lock file existing doesn't fail the mailserver running/functioning properly.

What do you think?

Reply:
Maybe we could check the lock file's age and if it's > an hour let's say, purge it. I think that'd be a good middle-ground.

Reply:
> as the lock file existing doesn't fail the mailserver running/functioning properly.

In that case, the lock file will never be deleted? Also updates (adding new users etc.) are not possible, while this container is running, right?

Maybe an ENV like "CLUSTERED_SETUP" might be thing? I don't know if it's worth. Are there other differences, where such a variable might be usefull now/in the future?

Reply:
> In that case, the lock file will never be deleted? Also updates (adding new users etc.) are not possible, while this container is running, right?

The user would have to manually delete it, yep. However, the lock file existing does not block executing any setup.sh and preparing the changes/changing the checksum. The lock file just prevents the check-for-changes script from running its logic and restarting dovecot & postfix.

> Maybe an ENV like "CLUSTERED_SETUP" might be thing? I don't know if it's worth. Are there other differences, where such a variable might be usefull now/in the future?

My initial response is against that due to the complexity it would add (we'd have to remember that it exists and double think our logic changes).

Reply:
I'm going to play around with something to clean up lock files over half an hour in age. This should minimize the damage IF the user orphans the lockfile somehow

Reply:
I appreciate it. But you decide if it's worth 😉  I just wanted to mention/talk about it and don't create exorbitant work for a small use case.